tornado/test/httpserver_test.py

```diff
@@ -423,18 +439,22 @@ def test_double_slash(self):
         self.assertEqual(200, response.code)
         self.assertEqual(json_decode(response.body), {})
 
-    def test_malformed_body(self):
-        # parse_qs is pretty forgiving, but it will fail on python 3
-        # if the data is not utf8.
-        with ExpectLog(gen_log, "Invalid x-www-form-urlencoded body"):
-            response = self.fetch(
-                "/echo",
-                method="POST",
-                headers={"Content-Type": "application/x-www-form-urlencoded"},
-                body=b"\xe9",
-            )
-        self.assertEqual(200, response.code)
-        self.assertEqual(b"{}", response.body)
+    def test_form_urlencode(self):
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+        uni_text = "chinese: \u5f20\u4e09"
+        for enc in ("utf8", "gbk"):
+            body = b"data=" + urllib.parse.quote(uni_text.encode(enc)).encode()
+            response = self.fetch("/" + enc, method="POST", headers=headers, body=body)
+            self.assertEqual(json_decode(response.body), {"echo": uni_text})
+
+    def test_form_raw(self):
+        # not technically correct, but commonly supported
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+        uni_text = "chinese: \u5f20\u4e09"
+        for enc in ("utf8", "gbk"):
+            body = b"data=" + uni_text.encode(enc)
```

Comment:
You *could* dedupe this with an extra loop `for body in [b"data="+uni_text.encode(enc), b"data="+urllib.parse.quote(uni_text.encode(enc)).encode()]`, but I'm not sure I would. It's often best not to get too clever in tests. 

We should probably also start using subtests for any loop-driven tests.

Reply:
I had not heard of `unittest.TestCase.subTest()`, I'll try it.